### PR TITLE
[SPARK-7788] Made KinesisReceiver.onStart() non-blocking

### PR DESCRIPTION
**Pull request by tdas**
KinesisReceiver calls worker.run() which is a blocking call (while loop) as per source code of kinesis-client library - https://github.com/awslabs/amazon-kinesis-client/blob/v1.2.1/src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java.
This results in infinite loop while calling sparkStreamingContext.stop(stopSparkContext = false, stopGracefully = true) perhaps because ReceiverTracker is never able to register the receiver (it's receiverInfo field is a empty map) causing it to be stuck in infinite loop while waiting for running flag to be set to false.

### CODE REVIEW

**Comment by tdas** [comment]
@cfregly please take a look

**Comment by tdas** [comment]
test this please.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 1]
```diff
@@ -126,8 +131,19 @@ private[kinesis] class KinesisReceiver(
...
+      override def run(): Unit = {
+        try {
+          worker.run()
+        } catch {
+          case NonFatal(e) =>
+            restart("Error running the KCL worker in Receiver", e)
+        }
+      }
+    }
+    workerThread.setName("Kinesis Receiver")
```
Comment (cfregly) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: is there a need to distinguish the thread names by receiver Id or some such?

**Comment by cfregly** [comment]
otherwise, looks good

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala [hunk 2]
```diff
@@ -126,8 +131,19 @@ private[kinesis] class KinesisReceiver(
...
     worker = new Worker(recordProcessorFactory, kinesisClientLibConfiguration)
-    worker.run()
-
+    workerThread = new Thread() {
+      override def run(): Unit = {
+        try {
+          worker.run()
+        } catch {
+          case NonFatal(e) =>
+            restart("Error running the KCL worker in Receiver", e)
```
Comment (zsxwing) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: Knowing nothing about Kinesis. Just a small question: if we call `worker.run` twice, what will happen? I'm not sure if `worker` is reusable.

Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: Its not. And the ReceiverSupevisor guarnatees (assumption) that onStart() is not going to be called twice, so this wont be reused.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala (hunk 1 above)
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala]: Prolly a good idea.

**Comment by tdas** [comment]
I am merging this because no Kinesis test is affected, and Jenkins has successfully compiled it.
Thanks guys for reviewing.

**Comment by cfregly** [comment]
looks good

**Pull request closed without merging**